cmd/mt-whisper-importer-writer/main.go

```diff
@@ -0,0 +1,289 @@
+	if err != nil {
+		panic(fmt.Sprintf("Error creating listener: %q", err))
+	}
+}
+
+func throwError(msg string) {
+	msg = fmt.Sprintf("%s\n", msg)
+	if *exitOnError {
+		panic(msg)
+	} else {
+		printLock.Lock()
+		fmt.Fprintln(os.Stderr, msg)
+		printLock.Unlock()
+	}
+}
+
+func log(msg string) {
+	if *verbose {
+		printLock.Lock()
+		fmt.Println(msg)
+		printLock.Unlock()
+	}
+}
+
+func (s *Server) chunksHandler(w http.ResponseWriter, req *http.Request) {
+	metric := &archive.Metric{}
+	err := metric.UnmarshalCompressed(req.Body)
+	if err != nil {
+		throwError(fmt.Sprintf("Error decoding metric stream: %q", err))
+		return
+	}
+	log("Handling new metric")
+
+	if len(metric.Archives) == 0 {
+		throwError("Metric has no archives")
+		return
+	}
+
+	avg := false
+	if whisper.AggregationMethod(metric.AggregationMethod) == whisper.AggregationAverage {
+		avg = true
+	}
+
+	partition, err := s.Partitioner.Partition(&metric.MetricData, int32(*numPartitions))
+	err = s.Index.AddOrUpdate(&metric.MetricData, partition)
+	if err != nil {
+		throwError(fmt.Sprintf("Error updating metric index: %q", err))
+		return
+	}
+
+	for archiveIdx, a := range metric.Archives {
+		archiveTTL := a.SecondsPerPoint * a.Points
+		tableTTL, err := s.selectTableByTTL(archiveTTL)
+		if err != nil {
+			throwError(fmt.Sprintf("Failed to select table for ttl %d in %+v: %q", archiveTTL, s.TTLTables, err))
+			return
+		}
+		entry, ok := s.TTLTables[tableTTL]
+		if !ok {
+			throwError(fmt.Sprintf("Failed to get selected table %d in %+v", tableTTL, s.TTLTables))
+			return
+		}
+		tableName := entry.Table
+
+		if !avg || archiveIdx == 0 || !*fakeAvgAggregates {
+			log(fmt.Sprintf(
+				"inserting %d chunks of archive %d with ttl %d into table %s with ttl %d and key %s",
+				len(a.Chunks), archiveIdx, archiveTTL, tableName, tableTTL, a.RowKey,
+			))
+			s.insertChunks(tableName, a.RowKey, tableTTL, a.Chunks)
+		} else {
+			// averaged archives are a special case because mt doesn't store them as such.
+			// mt reconstructs the averages on the fly from the sum and cnt series, so we need
+			// to generate these two series out of raw averaged data by multiplying each point
+			// with the aggregation span and storing the result as sum, cnt is the aggregation span.
+			type sumCntChunks struct {
+				sum *chunk.Chunk
+				cnt *chunk.Chunk
+			}
+
+			// seconds per point is assumed to be the aggregation span
+			aggSpan := a.SecondsPerPoint
+
+			sumArchive := make([]chunk.IterGen, 0, len(a.Chunks))
+			cntArchive := make([]chunk.IterGen, 0, len(a.Chunks))
+			for _, ig := range a.Chunks {
+				T0 := ig.Ts
+				sum := chunk.New(T0)
+				cnt := chunk.New(T0)
+
+				it, err := ig.Get()
+				if err != nil {
+					throwError(fmt.Sprintf("failed to get iterator from itergen: %q", err))
+					continue
+				}
+
+				for it.Next() {
+					ts, val := it.Values()
+					cnt.Push(ts, float64(aggSpan))
+					sum.Push(ts, val*float64(aggSpan))
```

Comment:
this doesen't look right to me. consider: 
original input data:

ts |  val
---|---
0 |  10
5 |12
10| 8
15 |10
20 |14
25 | 16
30 | 14
35 | 16

aggregated with avg ( 20 s per point):


ts |  val
---|---
15 |(10+12+8+10)/4 = 10
35 | (14+14+14+16)/4=15

according to the code, we would store for cnt:

ts |  val
---|---
15 | 20
35 | 20

and for sum:

ts |  val
---|---
15 | 200
35 | 300

while `sum/cnt` results in the correct avg, the sum and cnt themselves are not correct. Why don't we report instead the actual amount of points that were added together to go from the prior archive into this one, as counts, and for sum, the actual sums (i.e. the avg times the count)?
for cnt:

ts |  val
---|---
15 | 4
35 | 4

and for sum:

ts |  val
---|---
15 | 40
35 | 60